Bugzilla/User.pm

```diff
@@ -2754,6 +2755,127 @@ sub set_bounce_count {
 }
 
 
+###############
+# Phabricator #
+###############
+
+# Populate phabricator user info.
+sub ensure_phab {
+  my $self = shift;
+
+  if ($self->{phab}) {
+    # Already cached.
+    return 1;
+  }
+
+  if ($self->{phab_error}) {
+    # Already hit some error.
+    return 0;
+  }
+
+  (my $phab_url = Bugzilla->params->{phabricator_base_uri}) =~ s{/$}{};
+  my $phab_token = Bugzilla->params->{phabricator_api_key};
+  unless ($phab_url && $phab_token) {
+    $self->{phab_error} = "Phabricator is not set up";
+    return 0;
+  }
+
+  my $bmo_id = $self->id;
+
+  try {
+    my $ua = mojo_user_agent({request_timeout => 5});
+    $ua->transactor->name('BMO user profile');
+
+    # Map BMO id to Phabricator phid.
+    my $res = $ua->get(
+      "$phab_url/api/bmoexternalaccount.search" => form => {
+        'api.token' => $phab_token,
+        'accountids[0]' => $bmo_id});
+    my $ph_accounts = $res->result->json;
+
+    if (my $error = $ph_accounts->{error_info}) {
+      $self->{phab_error} = "Error in bmoexternalaccount.search: " . $error;
+      return 0;
+    }
+
+    if (scalar(@{$ph_accounts->{result}}) == 0) {
+      $self->{phab_error} = "No account found";
+      return 0;
+    }
+
+    my $phid = $ph_accounts->{result}[0]{phid};
+
+    # Map Phabricator phid to user name.
+    my $res = $ua->get(
+      "$phab_url/api/user.query" => form => {
+        'api.token' => $phab_token,
+        'phids[0]' => $phid});
+    my $ph_user = $res->result->json;
+
+    if (my $error = $ph_user->{error_info}) {
+      $self->{phab_error} = "Error in user.query: " . $error;
+      return 0;
+    }
+
+    if (scalar(@{$ph_user->{result}}) == 0) {
+      $self->{phab_error} = "No user found";
+      return 0;
+    }
+
+    my $user_name = $ph_user->{result}[0]{userName};
+    my $real_name = $ph_user->{result}[0]{realName};
+
+    $self->{phab} = {
+      phid => $phid,
+      user_name => $user_name,
+      real_name => $real_name,
+    };
+    return 1;
+  }
+  catch {
+    $self->{phab_error} = "Error: " . $_;
+    return 0;
+  };
+}
+
+sub phab_phid {
+  my $self = shift;
+  if (!$self->ensure_phab()) {
+      return "";
+  }
+
+  return $self->{phab}{phid};
+}
+
+sub phab_user_name {
+  my $self = shift;
+  if (!$self->ensure_phab()) {
+      return "";
+  }
+
+  return $self->{phab}{user_name};
+}
+
+sub phab_real_name {
+  my $self = shift;
+  if (!$self->ensure_phab()) {
+      return "";
+  }
+
+  return $self->{phab}{real_name};
+}
+
+# Returns error happened during ensure_phab if any.
+sub phab_error {
+  my $self = shift;
+  return $self->{phab_error};
+}
+
+sub phab_base_url {
+  (my $phab_url = Bugzilla->params->{phabricator_base_uri}) =~ s{/$}{};
+  return $phab_url;
+}
```

Comment:
I'm not sure if this is the right place.
let me know if there's better place, or if this info is already available in the template.